scRNA-seq-advanced/setup/integration-setup.Rmd

```diff
@@ -0,0 +1,402 @@
+
+  # Normalize:
+  qclust <- scran::quickCluster(filtered_sce)
+  filtered_sce <- scran::computeSumFactors(filtered_sce, clusters = qclust)
+  norm_sce <- scater::logNormCounts(filtered_sce)
+  
+  # Return the normalized SCE
+  return(norm_sce) 
+}
+
+# Note: I do think we should use `sce_list` throughout rather than different variable names because of _memory_
+sce_list <- purrr::map(
+  sce_list,
+  filter_and_normalize_sce
+)
+```
+
+
+
+## Combine the SCEs
+
+We now need to combine these SCEs into 1 object, which requires some bookkeeping.
+Specifically, we need to ensure:
+
+- Genes/features (aka, the `rowData` slot names) must match
+- The `colData` column names must match
+- That we can identify the original library each cell came from
+- Other useful information can still be found
+
+
+```{r}
+# What are the genes in common? Hold onto this vector!
+# For these data they already had the same genes, but no guarantees of that
+#   with data in the wild
+shared_genes <- sce_list %>%
+  # get rownames for each entry in sce_list
+  purrr::map(rownames) %>%
+  # find their intersection
+  purrr::reduce(intersect)
+
+# What are the colData names?
+purrr::map(sce_list, 
+           ~names(colData(.))) 
+
+# Some purrr magic can quickly show us any differences 
+purrr::map(sce_list, 
+           ~names(colData(.))) %>%
+  purrr::reduce(setdiff)
+
+# Let's have a peek at the rowData:
+# `mean` and `detected` are SCE-specific, but `gene_symbol` is not! Bookkeeping!
+purrr::map(sce_list, 
+           ~names(rowData(.)))
+```
+
+We want to format each SCE to meet our criteria:
+```{r}
+format_sce_list <- function(sce, sce_name, shared_genes) {
+  
+  # Reduce to shared genes
+  sce <- sce[shared_genes,]
+
+  # Update the colData row names so we can know where cells came from back
+  # Alternatively we can just get rid of a lot of this?
+  # glue is great!
+  rownames(colData(sce)) <- glue::glue("{sce_name}-{rownames(colData(sce))}")
+  
+  # Update the rowData names, except `gene_symbol`
+  names(rowData(sce)) <- ifelse(names(rowData(sce)) == "gene_symbol",
+                                "gene_symbol",
+                                glue::glue("{sce_name}-{rownames(rowData(sce))}")
+    
+  ) 
+  # Add in batch information as stand-alone column 
+  colData(sce)$batch <- sce_name
+  
+  # Return 
+  return(sce)
+}
+
+
+# Prepare the sce_list for combining
+sce_list <- purrr::imap(sce_list, 
+                        format_sce_list,
+                        shared_genes)
+
+# Combine!
+combined_sce <- do.call(cbind, sce_list)
+combined_sce
+
+# We can remove this now if we want for memory:
+rm(sce_list)
+```
+
+
+## Dimension reduction
+
+Before proceeding with integration, let's calculate UMAP embeddings on the unintegrated combined SCE.
+Then, we can compare the UMAP pre- and post-integration.
+
+```{r}
+# calculate high-variance genes to use for dimension reduction
+gene_variance <- scran::modelGeneVar(combined_sce)
```

Comment:
I just remembered that this is slightly different for merged datasets and should be 
```suggestion
gene_variance <- scran::modelGeneVar(combined_sce, block = "batch")
```